src/rgw/rgw_lua_background.cc

```diff
@@ -150,21 +144,31 @@ void Background::run() {
       std::unique_lock cond_lock(cond_mutex);
       cond.wait(cond_lock, [this]{return !paused || stopped;}); 
       if (stopped) {
+        // stopped while paused
         ldpp_dout(dpp, 10) << "Lua background thread stopped" << dendl;
         return;
       }
       ldpp_dout(dpp, 10) << "Lua background thread resumed" << dendl;
     }
-    const auto rc = read_script();
-    if (rc == -ENOENT || rc == -EAGAIN) {
-      // either no script or paused, nothing to do
-    } else if (rc < 0) {
-      ldpp_dout(dpp, 1) << "WARNING: failed to read background script. error " << rc << dendl;
-    } else {
+  
+    std::string script;
+    {
+      std::unique_lock cond_lock(pause_mutex);
+      if (!paused) {
+        set_package_path(L, lua_manager->get_luarocks_path());
+        std::string no_tenant;
+        const auto rc = rgw::lua::read_script(&dp, driver, no_tenant, 
+            null_yield, rgw::lua::context::background, script);
```

Comment:
why is it necessary to hold this lock over the call to `read_script()`?

looking at the lock above:
```c++
    if (paused) {
      ldpp_dout(dpp, 10) << "Lua background thread paused" << dendl;
      std::unique_lock cond_lock(cond_mutex);
      cond.wait(cond_lock, [this]{return !paused || stopped;}); 
```
`paused` isn't atomic - shouldn't we acquire the lock before checking it? if we take the lock there, we should be able to call `set_package_path()` under that same lock instead of reacquiring it here

Reply:
> why is it necessary to hold this lock over the call to `read_script()`?
> 

what if I'm reading the script (= RADOS object), and then real reload is triggered? 
I want the function that tries to call `Background::pause()` to be blocked until the read ends

> looking at the lock above:
> 
> ```c++
>     if (paused) {
>       ldpp_dout(dpp, 10) << "Lua background thread paused" << dendl;
>       std::unique_lock cond_lock(cond_mutex);
>       cond.wait(cond_lock, [this]{return !paused || stopped;}); 
> ```
> 
> `paused` isn't atomic - shouldn't we acquire the lock before checking it? if we take the lock there, we should be able to call `set_package_path()` under that same lock instead of reacquiring it here

`paused` is just a boolean, it is being set atomically (although there could be a race condition when one thread sets it and another one reads it but still doe not see the new value). 
this means that:
* `puase()` was called, but we still see "pause=false", but in this case, we would see the right value in line 157, won't read the script and won't run it. and just sleep until the next iteration
* `resume()` was called, but we still see "pause=true", in this case we would get into the "if" statement in line 142, and wait. ***agree there is an issue here***, we may get stuck on `wait` :-(

Reply:
> `paused` is just a boolean, it is being set atomically (although there could be a race condition when one thread sets it and another one reads it but still doe not see the new value).

there's really no need to be clever about this, just to avoid a lock every 5 seconds

and i still think this would be simpler if we took out this pause/resume logic and just stop/restart the background thread during realm reload

> what if I'm reading the script (= RADOS object), and then real reload is triggered?

this wouldn't be a problem if we shutdown/join the thread before destroying the rados store

Reply:
ok. agree. its is going to be simpler with paus/resume == shutdown/start

Reply:
fixed here: 0beafd639de33ca328badf1f849709fd41cde7ed

Reply:
@cbodley now background is explicitly shutdown+start in the realm reloader.
this also makes the lua background independent of RADOS